Tracks paste events in Calypso editor

nosolosw opened the pull request:
This PR adds the ability to track paste events in Calypso editor.

mtias commented on config/production.json:
```diff
@@ -69,6 +69,7 @@
 		"plans/personal-plan": true,
 		"post-editor/author-selector": true,
 		"post-editor/insert-menu": true,
+		"post-editor/track-paste-events": true,
```
Why are you adding a config flag if you are enabling in all environments?

nosolosw replied:
I may want to disable this on `desktop`, depending on how event tracking goes.

nosolosw replied:
To clarify on the above: still not sure how the clipboard API will play with an electron-based app, on several OS, so I've just prepared the code for the case we need to disable that.

mtias replied:
Fair enough. Seems like you could add to the _shared config file and disable on desktop if needed: https://github.com/Automattic/wp-calypso/blob/master/config/_shared.json

nosolosw replied:
Done.

nosolosw replied:
Gave it a second thought: if we need that, we can add it later. Remove flags for now.

nosolosw commented:
Note to self: modify the boolean `is_gdocs` for a more semantic `source` that would contain the list of identified sources as we list them: `unknow`, `google_docs`, etc.

dmsnell commented on client/components/tinymce/plugins/wpcom-paste-track/plugin.js:
```diff
@@ -0,0 +1,84 @@
+const SOURCE_GOOGLE_DOCS = 'google_docs';
+const SOURCE_UNKNOWN = 'unknown';
+
+function pasteTrack( editor ) {
```
the naming here is backwards. it reads as a noun. probably it should be a verb since it is there _to do something_.

`trackPaste` for example

dmsnell commented on client/components/tinymce/plugins/wpcom-paste-track/plugin.js:
```diff
@@ -0,0 +1,84 @@
+		debug( 'track paste event ' );
+		store.dispatch( recordTracksEvent( 'calypso_editor_content_paste', {
+			types: types,
+			mode: mode,
```
can go ahead and use the object property shorthand here

```js
{ types, mode, source: getSource( … ) }
```

(please ignore other formatting changes)

dmsnell commented on client/components/tinymce/plugins/wpcom-paste-track/plugin.js:
```diff
@@ -0,0 +1,84 @@
+		return types.reduce( ( a, b ) => {
+			return a + ', ' + b;
+		} );
+	}
```
what's the purpose of this function?

are we just wanting to join an array of strings with a comma?

nosolosw replied:
Yes, that's the purpose - but we need a fallback for cases where the `orgTypes` parameter is a [DOMStringList](https://developer.mozilla.org/en-US/docs/Web/API/DOMStringList).

dmsnell replied:
okay, so the purpose is to flatten arrays which might contain `DOMStringList`s. now, what is the purpose of this function more generally? why are we flattening the lists?

nosolosw replied:
The naming was weird, specifically `flatten` didn't express well what the code was trying to do. Changed the naming.

dmsnell commented on client/components/tinymce/plugins/wpcom-paste-track/plugin.js:
```diff
@@ -0,0 +1,84 @@
+	const textarea = editor.getParam( 'textarea' );
+	if ( textarea ) {
+		textarea.addEventListener( 'paste', onPasteTextArea );
+	}
```
do we need to pay attention to remove event listeners? they have a tendency to stick around longer than we expect.

dmsnell commented on client/components/tinymce/plugins/wpcom-paste-track/plugin.js:
```diff
@@ -0,0 +1,84 @@
+	}
+
+	function onPasteTextArea( event ) {
+		recordEvent( flattenArray( event.clipboardData.types ), 'html' );
```
`tinymce` seems to convey enough to indicate that the event originated there, but `html` seems fuzzier. `html` is also the content itself. maybe having something like, `html-editor` or `from-html` and `from-tinymce` or another iteration of naming could clarify

nosolosw replied:
`tinymce` and `html` are the modes available in TinyMCE editor, so I assumed people was familiar with the names. On a second thought, `editor-tinymce` and `editor-html` are more obvious and retain also the reference to the TinyMCE original modes for those familiar with them.

nosolosw replied:
Even better: for non-coders, `visual-editor` and `html-editor` will work arguably better, as this is the info we see in the editor tabs.

dmsnell commented on client/components/tinymce/plugins/wpcom-paste-track/plugin.js:
```diff
@@ -0,0 +1,84 @@
+		}
+		return types.reduce( ( a, b ) => {
+			return a + ', ' + b;
+		} );
```
won't it always be array-like with `Array.from()`? if that's the case, won't it be safe to `Array.from( listOfThings )` too?

```js
function flattenArray( list ) {
	return Array.from( list ).join( ', ' );
}
```

nosolosw replied:
this is what I just did, which is objectively simpler than the initial approach:

```js
function toString( types ) {
	return Array.from( types ).join( ', ' );
}
```

nosolosw commented:
@dmsnell this is ready for a second round.

dmsnell commented on client/components/tinymce/plugins/wpcom-track-paste/plugin.js:
```diff
@@ -46,39 +42,38 @@ function pasteTrack( editor ) {
+		// We need coverage for thoses cases, hence the Array.from conversion.
+		//
+		// https://html.spec.whatwg.org/multipage/interaction.html#datatransfer
+		// DOMStringListhttps://developer.mozilla.org/en-US/docs/Web/API/DOMStringList
```
this would be great as a JSDoc comment instead of an inline comment

also, this function doesn't depend on `this` so maybe we can move it out of the encapsulating function and into the module scope.

I like the direction of `toString` but I wonder if that's most appropriate since we already have `toString` as a standard function with standard expectations (it _might_ be the right one - I just want us to pause and make sure it is)

nosolosw replied:
Moved as a JSDoc comment. I don't have a strong opinion about the `toString` name, but I think it is OK to have it as it is, as it conveys a well know meaning.

dmsnell commented on client/components/tinymce/plugins/wpcom-track-paste/plugin.js:
```diff
@@ -46,39 +42,38 @@ function pasteTrack( editor ) {
-	function onPasteTextArea( event ) {
-		recordEvent( flattenArray( event.clipboardData.types ), 'html' );
+	function onPasteFromHTMLEditor( event ) {
+		recordEvent( 'html-editor', toString( event.clipboardData.types ) );
```
if we call `toString` on every call to `recordEvent` do we even need to call `toString` here? why not move it into `recordEvent` since `recordEvent` is already tightly coupled to recording these specific types of events.

could even make that clear by renaming to `recordPasteEvent` or something like that

dmsnell commented on client/components/tinymce/plugins/wpcom-track-paste/plugin.js:
```diff
@@ -46,39 +42,38 @@ function pasteTrack( editor ) {
 
-	// Bind paste event listeners to both Visual and HTML editors
-	editor.on( 'paste', onPasteEditor );
+	debug( 'bind listeners for paste event' );
```
we probably want to do a full review over which debug statements are valuable to us before we merge.

my guess is that many of these are more useful for the development period building this feature than they will be to leave in there after development has finished.

nosolosw replied:
Unless there is a guide in calypso that recommends deleting the `debug` statements, my personal preference is for maintain them as they are, as they are useful not only in the development phase, but in the debugging as well - when the worst happens, better to have the tools already in place.

dmsnell replied:
I won't stop you. just remember that every additional line of code is extra maintenance burden and every line of code not essential to the main thing is distracting us when we read it

:+1:

dmsnell commented on client/components/tinymce/plugins/wpcom-track-paste/plugin.js:
```diff
@@ -46,39 +42,38 @@ function pasteTrack( editor ) {
+		debug( 'unbind listeners for paste event on editor removal' );
+		editor.off( 'paste', onPasteFromTinyMCEEditor );
+		textarea.removeEventListener( 'paste', onPasteFromHTMLEditor );
+	} );
```
very nice!

nosolosw commented:
@dmsnell this is ready for the 3rd round. Commits squashed and branch rebased. Thanks for the review, I feel the code is in better shape now.

dmsnell commented on client/components/tinymce/index.jsx:
```diff
@@ -63,7 +64,8 @@ import EditorHtmlToolbar from 'post-editor/editor-html-toolbar';
 	toolbarPinPlugin,
-	embedReversalPlugin
+	embedReversalPlugin,
+	trackPaste
```
let's also add in a trailing comma here. helps keep things consistent and diffs smaller/less confusing

dmsnell commented on client/components/tinymce/index.jsx:
```diff
@@ -128,7 +130,8 @@ const PLUGINS = [
 	'wpcom/sourcecode',
-	'wpcom/embedreversal'
+	'wpcom/embedreversal',
+	'wpcom/trackpaste'
```
(trailing comma)

dmsnell commented on client/components/tinymce/plugins/wpcom-track-paste/plugin.js:
```diff
@@ -0,0 +1,92 @@
+	}
+
+	/**
+	 * Records the event.
```
"the event"

maybe something like _"Records paste events in the editor"_ could be a little more helpful when people find this function out of context.

nosolosw replied:
Sure.

dmsnell commented on client/components/tinymce/plugins/wpcom-track-paste/plugin.js:
```diff
@@ -0,0 +1,92 @@
+	 * Records the event.
+	 *
+	 * @param {String} mode - 'html-editor' or 'visual-editor', indicates which editor was in use on paste.
+	 * @param {(Array|DOMStringList)} types - The types the content is available to paste.
```
nice typing!

*(note that the `-` at the front of the description isn't necessary and it may end up actually a bit confusing for tools which interpret the JSDoc comments)

dmsnell commented on client/components/tinymce/plugins/wpcom-track-paste/plugin.js:
```diff
@@ -0,0 +1,92 @@
+	 * @param {(Array|DOMStringList)} types - The types the content is available to paste.
+	 */
+	function recordPasteEvent( mode, types ) {
+		debug( 'track paste event ' );
```
there's an unexpected space here.

dmsnell commented on client/components/tinymce/plugins/wpcom-track-paste/plugin.js:
```diff
@@ -0,0 +1,92 @@
+		const typesStr = toString( types );
+		store.dispatch( recordTracksEvent( 'calypso_editor_content_paste', {
+			mode,
+			types: typesStr,
```
why create the variable above? if we wanted we could inline it here…

```js
store.dispatch( …, {
	mode,
	types: toString( types ),
	source: …
```

dmsnell commented on client/components/tinymce/plugins/wpcom-track-paste/plugin.js:
```diff
@@ -0,0 +1,92 @@
+
+	function isGoogleDocs( types ) {
+		return ( types.indexOf( 'application/x-vnd.google-docs-image-clip+wrapped' ) > -1 ) ||
+		( types.indexOf( 'application/x-vnd.google-docs-document-slice-clip+wrapped' ) > -1 );
```
let's check the formatting here, also check out some array functions we can use to clarify what kind of operation we're intending on using here.

for example, are you familiar with the `intersection` operation? it pulls out the elements of two lists which are in both lists…

```js
import { intersection } from 'lodash';

const isGoogleDocs = types => intersection( types, [
	'application/x-vnd.google-docs-image-clip+wrapped',
	'application/x-vnd.google-docs-document-slice-clip+wrapped',
] ).length > 0;

// or…
import {
	flowRight as compose,
	intersection,
	isEmpty,
	negate,
	partial,
} from 'lodash';

const isGoogleDocs = types => compose(
	negate( isEmpty ),
	partial( intersection, [
		'application/x-vnd.google-docs-image-clip+wrapped',
		'application/x-vnd.google-docs-document-slice-clip+wrapped',
	] ),
);
```

nosolosw replied:
Pulling all these libraries for a simple operation like this seemed to me quite excessive. But I agree the code could be more readable, so I ended up refactoring it and converted the `isGoogleDocs( types ) ` param to an array (it was an string before), so we have a much more semantic code.

dmsnell commented on client/components/tinymce/plugins/wpcom-track-paste/plugin.js:
```diff
@@ -0,0 +1,92 @@
+		if ( isGoogleDocs( types ) ) {
+			return SOURCE_GOOGLE_DOCS;
+		}
+		return SOURCE_UNKNOWN;
```
this structure isn't needed if we only expect two types. we could instead return the value from a ternary expression…

```js
const getSource = types => isGoogleDocs( types )
	? SOURCE_GOOGLE_DOCS
	: SOURCE_UNKNOWN;
```

nosolosw replied:
I am expecting that we can discover more sources after this is deployed, so my mind kind of prepared the code for that fact. For now, it is simpler to use ternary operator and let's cross that bridge when/if we get there.

dmsnell commented on client/components/tinymce/plugins/wpcom-track-paste/plugin.js:
```diff
@@ -0,0 +1,92 @@
+	 * More info at:
+	 *
+	 * - types at https://html.spec.whatwg.org/multipage/interaction.html#datatransfer
+	 * - DOMStringList at https://developer.mozilla.org/en-US/docs/Web/API/DOMStringList
```
fun fact! we have a `@see` JSDoc command to reference external information

dmsnell commented:
@nosolosw thanks for following-up on the review. it's looking better indeed! we may disagree on the number of debug statements, which is fine. still it looks like a bit excessive to me given the risks involved in needing them. another round done!

nosolosw commented on client/components/tinymce/plugins/wpcom-track-paste/plugin.js:
```diff
@@ -0,0 +1,92 @@
+	const store = editor.getParam( 'redux_store' );
+	if ( ! store ) {
+		debug( 'store not found, do not bind events' );
+		return;
```
This is one thing I got inspiration from other plugins. From the perspective of `track-paste` plugin, I'm OK with let this as it is, as my assumption is that if Redux is missing there is a bigger problem - and my feeling is that it is the editor responsibility to fix it, not plugins.

dmsnell replied:
> if Redux is missing there is a bigger problem - and my feeling is that it is the editor responsibility to fix it, not plugins

in this case, why test to make sure that it's there?

the questions I would ask are as follows:
 - do we expect the store to be there sometimes and not always? if so, should this plugin simply not load if it's missing? if so, `return` is fine
 - do we require that the store be there always? if so, do we expect some other code to be responsible for loading it? if so, we should either throw (or maybe more appropriately) just _assume_ that it's there. if it's not there we will end up `throw`ing an error anyway when we try to dispatch an action.

personally I would be in favor of removing this code: it's redundant if we really believe that the store should be here; it's confusing because it leads to the question "when would this happen and what should happen?"

nosolosw replied:
You have a valid point here - code removed.

nosolosw commented:
@dmsnell I'm addressed all your suggestions and also went ahead and refactor the code to be more `es6` friendly, which seemed to me that was the spirit behind some of your comments, provides better readability and I see now that it is the current calypso practice.

I agree that having `debug` statements means more code to maintain, but I would call that a tradeoff, not a risk - risk is a word with a strong meaning that I reserve for security issues and the like. Said that, I'm happy to remove them if that's the current calypso practice.

Let me know if there is some blocker/thing to improve left or we can merge this.

dmsnell commented on client/components/tinymce/plugins/wpcom-track-paste/plugin.js:
```diff
@@ -19,58 +23,35 @@ function trackPaste( editor ) {
-		return ( types.indexOf( 'application/x-vnd.google-docs-image-clip+wrapped' ) > -1 ) ||
-		( types.indexOf( 'application/x-vnd.google-docs-document-slice-clip+wrapped' ) > -1 );
-	}
+	const isGoogleDocs = types => types.filter( isGoogleDocsType ).length > 1;
```
there's a built-in array function `.some()` which would accomplish the same result but also indicate to others that our goal is to test if any of the items in the array match the given predicate

nosolosw replied:
Yup, I'd rather use `some`, but [my understanding](https://github.com/Automattic/wp-calypso/pull/10812#discussion_r97765523) is that we cannot use it as that transformation is provided by the `babel-polyfill`, one babel plugin that we don't have.

dmsnell replied:
`Array.some()` goes back to IE 9 and is in every other browser. Are we still supporting IE 8 @aduth?

if that's the case it's very reasonable to pull in `lodash.some()` which doesn't have these issues.

```js
some( types, isGoogleDocs )
```

nosolosw replied:
You're right, I misread the specification table for `some` and thought it was `es2015`. We seem to support IE11 and edge, so I'll go ahead and update the code to use `some` instead.

nosolosw replied:
Deployed and tested.

dmsnell commented:
@nosolosw even though I left another comment, I think this is okay to merge. I would appreciate it if you considered the comment I left though.

further, I'm still not a big fan of those debug statements. I think part of it is that they are doing nothing but stating what is plain in the code that follows. in other words, they don't indicate runtime events which may or may not occur. they solely state events that will always happen. the one inside of the paste handler is different because that describes at runtime events which we cannot see beforehand, but it's odd to log lines of code. maybe the `init` message is another _more-valid_ case.

at some level there have been subjective evaluations on when to use them since we don't have a debug statement in front of every line of code :wink:. I won't bring it up again as I trust you to make your own decisions on this - there is no guide on it.

nosolosw commented:
@dmsnell I've been revisiting the debug statements and removed some that I consider now dispensable.

Also squashed the commits.

dmsnell commented:
looks ready! how about some tests before the merge? I would be glad to help out if that seems scary or unfamiliar to you.

nosolosw commented:
Will love that! I'm going to merge this first, as we want some early data, and will follow up with a new PR for the tests.

dmsnell commented:
> Will love that! I'm going to merge this first, as we want some early data, and will follow up with a new PR for the tests.

next time let's make the tests _before_ merging :wink:

if we think about the tests _while_ or even _before_ we develop they will inform the way we code and we will catch bugs before we push out to production.

aduth commented:
Speaking of bugs in production, the timing of merge plus contents of changes leads me to believe this is related to a recent spike in errors on the `/post/:siteId` path:

>Unable to get property 'types' of undefined or null reference

Will message directly with more details.

nosolosw commented:
Couldn't reproduce, but this https://github.com/Automattic/wp-calypso/pull/11463 may address the issue.
